keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,617 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP extends the Job API to allow setting conditions under which an Indexed Job can be declared as succeeded.
+
+## Motivation
+
+There are cases where a batch workload requires an indexed job that want to care 
+only about leader indexes in determining the success or failure of a Job,
+for example MPI and PyTorch etc. This is currently not possible because the indexed job 
+is marked as Completed only if all indexes are succeeded.
+
+Additionally, there are cases where a batch workload can be declared as succeeded if a number of
+pods succeed, regardless of the index. Sometimes the number needs
+because an indexed job author might not know how many indexes the job will have in advance
+when the above batch workload has an autoscale semantic.
+
+Some third-party frameworks have implemented success policy.
+
+- [Kubeflow Training Operator](https://www.kubeflow.org/docs/components/training)
+- [Flux Operator](https://flux-framework.org/flux-operator/)
+
+### Goals
+
+- allow to mark a job as a `Succeeded` even if some of the indexes fail.
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` mode.
+- Change the existing behavior of `Indexed` mode without a SuccessPolicy.
+
+## Proposal
+
+We propose new policies under which a job can be declared as succeeded.
+Those policies can be modeled in the following:
+
+1. An indexed job completes if a set of [x, y, z...] indexes are successful.
+2. An indexed job completes if x or x% of indexes are successful.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine-learning researcher, I run an indexed job which leaders are launched
+as index=0,1,2, and workers are launched as index=3+.
+I want to care about only the leaders when the result of job is evaluated.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    onSucceededPods:
+      indexes: "0-2"
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a machine-learning researcher, I run an indexed job with an autoscale semantic
```

Comment:
I assumed to use horovod.

Reply:
I know horovod supports a form of autoscaling, but does horovod actually have those success semantics? if so, please add to the KEP and a reference to Horovod docs.

IIUC, the Horovod case should be addressed by mutating the .spec.completions to scale-down the job (the ElasticIndexedJob feature), and so the success semantics doesn't change: all pods should succeed, which in this case would be the reduced completion count.

The reason I am asking for this is because some of the stories mentioned here seem hypothetical and lacking a real-world example.

Reply:
It sounds reasonable. Thank you for pointing out this :)

I reinvestigated PyTorch+HPA (implemented in kubeflow), MPI+Horovod, and PyTorch+Horovod with autoscaling. As a result, I find that all cases require that all worker ranks (MPIRank/PyTorchRank) succeed, and frameworks don't support multiple leaders, although they allow them to recreate ranks.

So, I think we should remove story 2 and story 3.
Are there any other concerns about autoscaling semantics?

Note that I think that `count: 5` and `count: 50%` forms are still effective for story 4.

Reply:
Done.